Optimize the lock which in the RunPredicate

godliness opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. If this is your first time, read our contributor guidelines https://git.k8s.io/community/contributors/guide#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
3. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/guide/release-notes.md
4. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

**What this PR does / why we need it**:


Enhance the performance of scheduler

-  Change the lock in the RunPredicate from lock to rlock




**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:

Could solve part of #63784 

**Special notes for your reviewer**:

_Run benchmark test by scheduler_perf_:
`Before`  BenchmarkScheduling/1000Nodes/0Pods-32     1000    11689758 ns/op
`After`     BenchmarkScheduling/1000Nodes/0Pods-32     1000    5951510 ns/op

_Run integration (density) test by scheduler_perf_:
Schedule 3000 Pods On 3000 Nodes
`Before`  rate  19 per second on average
`After`     rate  58 per second on average

_Cpu profile test result_:
`Before`  [click](https://cdn.rawgit.com/godliness/files/master/63784_before.svg)
`After`     [click](https://cdn.rawgit.com/godliness/files/master/63784_after.svg)

**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
`None`
```

/sig scheduling

/cc @misterikkit 
/cc @bsalamat
/cc @ravisantoshgudimetla 
/cc @resouer

godliness committed:
Optimize the lock which in the RunPredicate

resouer commented on pkg/scheduler/core/equivalence_cache.go:
```diff
@@ -71,9 +71,9 @@ func (ec *EquivalenceCache) RunPredicate(
-	defer ec.mu.Unlock()
+	ec.mu.RLock()
 	fit, reasons, invalid := ec.lookupResult(pod.GetName(), nodeInfo.Node().GetName(), predicateKey, equivClassInfo.hash)
+	ec.mu.RUnlock()
```
Please move RLock()/RUnlock() inside `lookupResult ` and `defer ec.mu.RUnlock()`

godliness replied:
thx, i will fix it.

resouer commented on pkg/scheduler/core/equivalence_cache.go:
```diff
@@ -83,7 +83,9 @@ func (ec *EquivalenceCache) RunPredicate(
 	if cache != nil && cache.IsUpToDate(nodeInfo) {
+		ec.mu.Lock()
 		ec.updateResult(pod.GetName(), nodeInfo.Node().GetName(), predicateKey, fit, reasons, equivClassInfo.hash)
+		ec.mu.Unlock()
```
dito, make the function itself thread safe is better

misterikkit replied:
Actually, we need to hold the lock when we call cache.IsUpToDate. This is what fixed the race condition in https://github.com/kubernetes/kubernetes/pull/63040. I think the window for errors is extremely small with this setup though.

godliness replied:
ok, i saw that, i will keep the lock  outside of cache.IsUpToDate.

resouer commented:
/ok-to-test

godliness commented:
/retest

godliness commented:
/retest

godliness commented:
I've updated the test result depends on the new code.

ravisantoshgudimetla commented:
Thanks @godliness for the PR and for the flame graphs. After I looked at them, I decided to run down some more tests to get more data.

Attaching the call graphs with and without cache changes(from this PR, I will add one before and after this PR too). 
- O.S: fedora 27 
- Cores: 8 core
- Golang: 1.10.2
- Source code pulled from latest master branch
- Scheduled: 10k fake pods on 3k fake nodes.

[cpu.pdf](https://github.com/kubernetes/kubernetes/files/2015391/cpu.pdf)
[cpu_withecache.pdf](https://github.com/kubernetes/kubernetes/files/2015392/cpu_withecache.pdf)

I will look more into graphs tomorrow but following is my preliminary analysis:
- While ecache is enabled we are spending around 50.35s (8.52%)of 273.73s (46.30%) in `lookUpResult` fn.
- We are also spending quite a lot of time in gc.
- In total for locking and unlocking we are spending around 47.96sec(for RLock).

Please feel free to add your observations and analysis too.

godliness commented:
/retest

godliness commented:
/retest

resouer commented on pkg/scheduler/core/equivalence_cache.go:
```diff
@@ -81,10 +79,12 @@ func (ec *EquivalenceCache) RunPredicate(
 	if cache != nil && cache.IsUpToDate(nodeInfo) {
 		ec.updateResult(pod.GetName(), nodeInfo.Node().GetName(), predicateKey, fit, reasons, equivClassInfo.hash)
 	}
+	ec.mu.Unlock()
```
Why don't move the lock part into`ec.updateResult()`?

Also, please update `equivalence_cache_test.go`.

godliness replied:
updated

godliness commented:
/retest

godliness commented:
/test pull-kubernetes-e2e-gce

godliness commented:
Why the pull-kubernetes-e2e-gce always failed?  How should i do for this?

godliness commented:
@ravisantoshgudimetla  Thanks for the detail test information.

From the "cpu_withecache" result:

- Looks like we have already used Rlock in the lookupResult fn

-  runtime.convT2Estring cost time in lookupResult, this is due to we used glog.v(string format) in this fn. and this lead to a lot of time in gc?   

From now on, i just could share this information.

godliness commented:
/test pull-kubernetes-e2e-gce

resouer commented:
RLock seems solve small part of the problem but not enough. Now I guess the major impact is caused by the lock to protect ecache from race of `factory.go` (need some tests to prove it). If that's the case, we can probably use other approach to replace that lock.

bsalamat commented on pkg/scheduler/core/equivalence_cache_test.go:
```diff
@@ -481,13 +481,13 @@ func TestLookupResult(t *testing.T) {
 		}
 		// calculate predicate with equivalence cache
-		ecache.mu.Lock()
+		ecache.mu.RLock()
```
You are double locking here. lookupResult has a lock now.

resouer commented:
According to my local test, this patch improved operation time a bit:

```
When ecache is enabled:

Original:

BenchmarkScheduling/1000Nodes/1000Pods-32        	     100	  54653471 ns/op	 3915357 B/op	   85775 allocs/op
BenchmarkSchedulingAntiAffinity/500Nodes/5000Pods-32        	     250	  20053089 ns/op	 2846708 B/op	   56706 allocs/op
BenchmarkScheduling/1000Nodes/0Pods-32           	     100	  55889442 ns/op	 3290233 B/op	   73855 allocs/op

With patch:

BenchmarkScheduling/1000Nodes/1000Pods-32        	     100	  50702524 ns/op	 3912885 B/op	   85740 allocs/op
BenchmarkSchedulingAntiAffinity/500Nodes/5000Pods-32        	     250	  18336367 ns/op	 2707880 B/op	   54136 allocs/op
BenchmarkScheduling/1000Nodes/0Pods-32           	     100	  54861638 ns/op	 3224870 B/op	   72515 allocs/op
```

resouer commented:
Emm ... op time is increased after patch in your test? 
> Before BenchmarkScheduling/1000Nodes/0Pods-32 1000 5995726 ns/op
After BenchmarkScheduling/1000Nodes/0Pods-32 1000 26993580 ns/op

godliness commented:
@resouer cause i added time.StopTimer() in line https://github.com/kubernetes/kubernetes/blob/master/test/integration/scheduler_perf/scheduler_bench_test.go#L160
 I think we don't need calculate the prepare time in the benchmark test, am i right?

godliness commented:
@resouer  I tested again in my local test

Because i added b.StopTimer() in line https://github.com/kubernetes/kubernetes/blob/master/test/integration/scheduler_perf/scheduler_bench_test.go#L160
I think we don't need calculate the prepare time in the benchmark test, am i right?

Don't Added b.StopTimer()
```
Before  BenchmarkScheduling/1000Nodes/0Pods-32  1000   29348329 ns/op
After  BenchmarkScheduling/1000Nodes/0Pods-32    1000    26033742 ns/op
```

Added b.StopTimer()
```
Before BenchmarkScheduling/1000Nodes/0Pods-32  1000   11689758 ns/op
After BenchmarkScheduling/1000Nodes/0Pods-32    1000    5951510 ns/op
```

By the way, I need reversed result between "before" and "after", and fill the after result with "Added b.StopTimer()"

godliness commented:
/retest

godliness commented:
/test pull-kubernetes-e2e-gce

resouer commented:
Sorry I didn't look it deeply but I'm not quite follow why you need to hack the test right now. While before we can move on, it's important that we should have the same baseline in this discussion ...

godliness commented:
@resouer 
I referenced this issue https://github.com/kubernetes/kubernetes/issues/61582
Please feel free to have a look. :-)

godliness commented:
/test pull-kubernetes-e2e-gce

bsalamat reviewed (commented):
@misterikkit Could you please take a look as well?

godliness commented on pkg/scheduler/core/equivalence_cache.go:
```diff
@@ -88,10 +85,12 @@ func (ec *EquivalenceCache) RunPredicate(
 	if err != nil {
 		return fit, reasons, err
 	}
+	ec.mu.Lock()
```
Thanks for this review, i have fixed it depends on your advice, hope could be OK.

godliness commented:
/rebase

resouer reviewed (commented):
The change generally looks good to me now. @misterikkit  Please take a final look?

@godliness Could you please post perf test result with this patch in your physical machine (and `NodeLabelCount=1`)? So we can know the optimization space we have here.

bsalamat commented on pkg/scheduler/core/equivalence_cache.go:
```diff
@@ -88,20 +85,28 @@ func (ec *EquivalenceCache) RunPredicate(
+	if !cache.IsUpToDate(nodeInfo) {
+		return
+	}
+	nodeName := nodeInfo.Node().GetName()
```
Please add a check to ensure that nodeInfo.Node() is not nil before dereferencing it. It could be nil in tests.

bsalamat replied:
The caller is already checking that, but if this function is used from another place, the new caller may not check. So, it is better to check in the function.

godliness replied:
Yes, you are right. 
Fixed it, thanks!

godliness commented:
@resouer  Please have a look :-)

**Cpus**: 32
**Mem**: 64GB
**Uame -a** :Linux LOCAL 3.10.0-327.28.3.el7.x86_64 #1 SMP Thu Aug 18 19:05:49 UTC 2016 x86_64 x86_64 x86_64 GNU/Linux
**Code Base**: master branch (9c2e51f37c5174d22d0d3819baf5ba4f6a9b2751) with or without this patch
**Golang**: 1.10.2


**Before （Original）**

```
NodeAffinity LabelCount=1
I0524 10:02:25.572008   41645 feature_gate.go:194] feature gates: map[EnableEquivalenceClassCache:true]
I0524 10:02:25.572641   41645 factory.go:1066] Created equivalence class cache
0s    rate: 847    total: 847 (qps frequency: map[847:1])
1s    rate: 599    total: 1446 (qps frequency: map[847:1 599:1])
2s    rate: 524    total: 1970 (qps frequency: map[847:1 599:1 524:1])
3s    rate: 483    total: 2453 (qps frequency: map[847:1 599:1 524:1 483:1])
4s    rate: 476    total: 2929 (qps frequency: map[847:1 599:1 524:1 483:1 476:1])
Scheduled 3000 Pods in 5 seconds (600 per second on average). min QPS was 476
I0524 10:02:33.581535   41645 util.go:83] destroying scheduler
I0524 10:02:33.581607   41645 util.go:87] destroyed scheduler
I0524 10:02:33.581624   41645 util.go:51] destroying API server
I0524 10:02:33.601822   41645 util.go:53] destroyed API server
Minimal observed throughput for 3k pod test: 476
PASS
ok      k8s.io/kubernetes/test/integration/scheduler_perf    11.925s
+++ [0524 10:02:33] ...density tests finished
/export/workspace/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/integration/scheduler_perf
+++ [0524 10:02:33] performance test cleanup complete
```

**After (With this patch)**

```
NodeAffinity LabelCount=1
I0524 10:03:44.896844   42385 feature_gate.go:194] feature gates: map[EnableEquivalenceClassCache:true]
I0524 10:03:44.897494   42385 factory.go:1066] Created equivalence class cache
0s    rate: 1416    total: 1416 (qps frequency: map[1416:1])
1s    rate: 1148    total: 2564 (qps frequency: map[1416:1 1148:1])
Scheduled 3000 Pods in 2 seconds (1500 per second on average). min QPS was 1148
I0524 10:03:49.881567   42385 util.go:83] destroying scheduler
I0524 10:03:49.881653   42385 util.go:87] destroyed scheduler
I0524 10:03:49.881678   42385 util.go:51] destroying API server
I0524 10:03:49.883425   42385 util.go:53] destroyed API server
Minimal observed throughput for 3k pod test: 1148
PASS
ok      k8s.io/kubernetes/test/integration/scheduler_perf    8.694s
+++ [0524 10:03:49] ...density tests finished
/export/workspace/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/integration/scheduler_perf
+++ [0524 10:03:50] performance test cleanup complete
```

**Without any hack to Benchmark Test**

**Before**

`BenchmarkScheduling/100Nodes/0Pods-32                 1000       5878945 ns/op`

**After**

`BenchmarkScheduling/100Nodes/0Pods-32                 1000       5102698 ns/op`

**With hack to Benchmark Test**

Why i hack the test?

Because we  do not calculate the time of making nodes and destroy them which really cost plenty time on it, and makes the benchmark test not accurate。 If i am wrong, please figure it out. I will really appreciate about that.  

How i hack the test?

Referenced this issue #61582

**Before**

`BenchmarkScheduling/100Nodes/0Pods-32                 1000       2137814 ns/op`

**After** 

`BenchmarkScheduling/100Nodes/0Pods-32                 1000       1219657 ns/op`

bsalamat reviewed (commented):
/lgtm

Thanks, @godliness!

godliness commented:
/test pull-kubernetes-e2e-gce
